### PR TITLE
feat: Integrate JSX into snap notifications

### PR DESCRIPTION
**Pull request by hmalik88**
## **Description**

Adds an expanded view for snaps notifications, using JSX content returned from the snap to populate the expanded view.

Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change? Allow for richer snaps notifications
2. What is the improvement/solution? Add expanded view for snaps, allowing a snap to return jsx content in the expanded view.

### **After**

https://github.com/user-attachments/assets/74c89bb7-7510-4d1c-acb6-a585978ecec8

## **Manual testing steps**
1. Pull down https://github.com/hmalik88/swiss-knife-snap and run `yarn start` to serve the snap and site.
2. Build this branch using `yarn start:flask`
3. Install the snap from the local host site.
4. Trigger a notification with the "trigger a JSX notification" button and observe the results.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/notifications/notification-components/index.ts [hunk 1]
```diff
@@ -9,9 +9,20 @@ import { components as StakeComponents } from './stake/stake';
...
+}
+
+export const TRIGGER_TYPES = {
+  ...NotificationServicesController.Constants.TRIGGER_TYPES,
+  ...SNAP_TRIGGER,
+};
+
+export type TRIGGER_TYPES =
+  | NotificationServicesController.Constants.TRIGGER_TYPES
+  | SNAP_TRIGGER;
```
Comment (hmalik88) [inline ui/pages/notifications/notification-components/index.ts:25]: Adjusting types here to accommodate for snaps, these will be adjusted upstream in a later PR.

**Comment by Mrtenz** [comment]
@metamaskbot update-policies

File: ui/components/multichain/notification-detail-button/notification-detail-button.tsx [hunk 2]
```diff
@@ -45,20 +58,25 @@ export const NotificationDetailButton = ({
...
+    if (notification.type === TRIGGER_TYPES.SNAP && isMetaMaskUrl) {
+      navigate(href);
+    }
   };
 
   return (
     <Button
       key={id}
-      href={href}
+      {...(!isMetaMaskUrl && { href })}
```
Comment (Mrtenz) [inline ui/components/multichain/notification-detail-button/notification-detail-button.tsx:77]: Does this work?
```suggestion
      href={!isMetaMaskUrl && href}
```

File: ui/contexts/metamask-notifications/metamask-notifications.tsx [hunk 3]
```diff
@@ -7,8 +7,11 @@ import { selectIsProfileSyncingEnabled } from '../../selectors/metamask-notifica
 import { selectIsMetamaskNotificationsEnabled } from '../../selectors/metamask-notifications/metamask-notifications';
 import { getUseExternalServices } from '../../selectors';
 import { getIsUnlocked } from '../../ducks/metamask/metamask';
+import { SnapNotification } from '../../pages/notifications/snap/types/types';
 
-type Notification = NotificationServicesController.Types.INotification;
+type Notification =
```
Comment (Mrtenz) [inline ui/contexts/metamask-notifications/metamask-notifications.tsx:12]: Since this is used in several places, maybe extract it to a separate file.

File: ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx [hunk 4]
```diff
@@ -39,6 +39,11 @@ export const NotificationDetailsFooter = ({
           <footer.ExtensionLink notification={notification} />
         </>
       )}
+      {footer.type === 'footer_snap_notification' && (
+        <>
```
Comment (Mrtenz) [inline ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx:43]: This fragment is unnecessary.

File: ui/pages/notifications/notification-components/index.ts [hunk 5]
```diff
@@ -9,9 +9,20 @@ import { components as StakeComponents } from './stake/stake';
 import { components as SwapCompletedComponents } from './swap-completed/swap-completed';
 import { components as LidoWithdrawalRequestedComponents } from './lido-withdrawal-requested/lido-withdrawal-requested';
 import { components as LidoStakeReadyToBeWithdrawnComponents } from './lido-stake-ready-to-be-withdrawn/lido-stake-ready-to-be-withdrawn';
+import { components as SnapNotificationComponents } from './snap/snap';
 
-const { TRIGGER_TYPES } = NotificationServicesController.Constants;
-type TRIGGER_TYPES = NotificationServicesController.Constants.TRIGGER_TYPES;
+export enum SNAP_TRIGGER {
+  SNAP = 'snap',
```
Comment (Mrtenz) [inline ui/pages/notifications/notification-components/index.ts:15]: Should this be in the notification controller instead? 🤔

File: ui/pages/notifications/notification-components/snap/snap.tsx [hunk 6]
```diff
@@ -6,68 +6,93 @@ import {
...
+        category: MetaMetricsEventCategory.NotificationInteraction,
+        event: MetaMetricsEventName.NotificationClicked,
+        properties: {
+          notification_id: notification.id,
+          notification_type: notification.type,
+          previously_read: notification.isRead,
+        },
+      });
+      history.push(getSnapRoute(notification.data.origin));
+    };
```
Comment (Mrtenz) [inline ui/pages/notifications/notification-components/snap/snap.tsx:46]: Nit:
```suggestion
    };

```

File: ui/pages/notifications/notification-components/snap/snap.tsx [hunk 7]
```diff
@@ -6,68 +6,93 @@ import {
...
+        snapMessage={notification.data.message}
+        handleSnapClick={onClick}
+        handleSnapButton={handleSnapButton}
+      />
+    );
+  },
+  details: {
+    title: ({ notification }) => (
+      <NotificationDetailTitle
+        title={notification.data.expandedView?.title as string}
```
Comment (Mrtenz) [inline ui/pages/notifications/notification-components/snap/snap.tsx:69]: Is this property guaranteed to be set?

File: ui/pages/notifications/notification-components/snap/snap.tsx [hunk 8]
```diff
@@ -6,68 +6,93 @@ import {
...
+    return (
+      <NotificationListItemSnap
+        id={notification.id}
+        snapId={notification.data.origin}
+        isRead={notification.isRead}
+        createdAt={new Date(notification.createdAt)}
+        title={{
+          items: [
+            {
+              text: snapsNameGetter(notification.data.origin) || 'Snap',
```
Comment (Mrtenz) [inline ui/pages/notifications/notification-components/snap/snap.tsx:56]: Maybe use "Unknown Snap"? Is there a case in which this happens in the first place?
```suggestion
              text: snapsNameGetter(notification.data.origin) ?? 'Unknown Snap',
```

File: ui/pages/notifications/notification-components/snap/snap.tsx [hunk 9]
```diff
@@ -6,68 +6,93 @@ import {
...
+        title={notification.data.expandedView?.title as string}
+        date={formatIsoDateString(notification.createdAt)}
+      />
+    ),
+    body: {
+      type: 'body_snap_notification',
+      Content: ({ notification }) => (
+        <SnapUIRenderer
+          snapId={notification.data.origin}
+          interfaceId={notification.data.expandedView?.interfaceId as string}
```
Comment (Mrtenz) [inline ui/pages/notifications/notification-components/snap/snap.tsx:78]: Is this property guaranteed to be set?

File: ui/pages/notifications/notification-components/snap/snap.tsx [hunk 10]
```diff
@@ -6,68 +6,93 @@ import {
...
+    },
+  },
+  footer: {
+    type: 'footer_snap_notification',
+    Link: ({ notification }) =>
+      notification.data.expandedView?.footerLink ? (
+        <NotificationDetailButton
+          notification={notification}
+          text={notification.data.expandedView?.footerLink.text as string}
+          href={notification.data.expandedView?.footerLink.href as string}
```
Comment (Mrtenz) [inline ui/pages/notifications/notification-components/snap/snap.tsx:92]: Are these properties guaranteed to be set?

File: ui/pages/notifications/notification-components/types/notifications/notifications.ts [hunk 11]
```diff
@@ -32,6 +35,11 @@ type BodyFeatureAnnouncement<N = Notification> = {
   Description: NotificationFC<N>;
 };
 
+type BodySnapNotification<N = Notification> = {
+  type: 'body_snap_notification';
```
Comment (Mrtenz) [inline ui/pages/notifications/notification-components/types/notifications/notifications.ts:39]: Can you extract these into a separate type / enum?

File: ui/pages/notifications/snap/types/types.ts [hunk 12]
```diff
@@ -1,8 +1,15 @@
-export const SNAP = 'SNAP' as const;
+import { SNAP_TRIGGER } from '../../notification-components';
+
+export type SnapNotificationExpandedView = {
```
Comment (Mrtenz) [inline ui/pages/notifications/snap/types/types.ts:3]: Maybe this should be a type in `snaps-sdk`, so we can re-use it.

File: app/scripts/metamask-controller.js [hunk 13]
```diff
@@ -2756,14 +2756,19 @@ export default class MetamaskController extends EventEmitter {
               origin,
               args.message,
             ),
-          showInAppNotification: (origin, args) =>
+          showInAppNotification: (origin, args) => {
+            const notificationArgs = { ...args };
+            notificationArgs.interfaceId = notificationArgs.content;
+            delete notificationArgs.content;
+            delete notificationArgs.type;
```
Comment (Mrtenz) [inline app/scripts/metamask-controller.js:2763]: What's going on here?

**Review by Mrtenz — DISMISSED** [review]

File: ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx (hunk 4 above)
Comment (hmalik88) [reply ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx:43]: I thought so too, was just following the notification team's pattern 😅, I'll change it.

File: ui/pages/notifications/notification-components/index.ts (hunk 5 above)
Comment (hmalik88) [reply ui/pages/notifications/notification-components/index.ts:15]: yeahh...but opening a PR in their repo would take longer. Doing this as a temp workaround.

File: ui/pages/notifications/notification-components/snap/snap.tsx (hunk 7 above)
Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:69]: yes, this is only ever called when the notification route is hit

File: ui/pages/notifications/notification-components/snap/snap.tsx (hunk 8 above)
Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:56]: None that I know of, i think the notifications team just had it as a fail-safe. I'll change it.

File: ui/pages/notifications/notification-components/snap/snap.tsx (hunk 9 above)
Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:78]: yup, see above.

File: ui/pages/notifications/notification-components/snap/snap.tsx (hunk 10 above)
Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:92]: Yes, see above.

File: ui/pages/notifications/snap/types/types.ts (hunk 12 above)
Comment (hmalik88) [reply ui/pages/notifications/snap/types/types.ts:3]: I can open an issue to add it in after and replace this.

File: app/scripts/metamask-controller.js (hunk 13 above)
Comment (hmalik88) [reply app/scripts/metamask-controller.js:2763]: Changing the shape to fit what the notification controller is expecting.

Comment (Mrtenz) [reply app/scripts/metamask-controller.js:2763]: Can you just create a new object, rather than spreading and deleting properties afterwards?

File: ui/pages/notifications/notification-components/snap/snap.tsx (hunk 7 above)
Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:73]: it is not

Comment (Mrtenz) [reply ui/pages/notifications/notification-components/snap/snap.tsx:73]: Maybe I misunderstand, but it's optional here:

https://github.com/MetaMask/snaps/blob/f4890dd4eaa1b318f6fad00aa456e4600bc97225/packages/snaps-sdk/src/types/methods/notify.ts#L40-L43

Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:73]: This component is only ever called in a notification detail route, the route is not navigated to unless there's a detailed view. The title is guaranteed to be there.

File: app/scripts/metamask-controller.js [hunk 14]
```diff
@@ -2792,7 +2800,7 @@ export default class MetamaskController extends EventEmitter {
 
             return this.controllerMessenger.call(
               'PhishingController:testOrigin',
-              sender.url,
+              sender,
```
Comment (FrederikBolding) [inline app/scripts/metamask-controller.js:2803]: ?

File: ui/components/multichain/create-eth-account/create-eth-account.js [hunk 15]
```diff
@@ -7,7 +7,7 @@ import {
   setAccountLabel,
   getNextAvailableAccountName as getNextAvailableAccountNameFromController,
 } from '../../../store/actions';
-import { CreateAccount } from '..';
+import { CreateAccount } from '../create-account';
```
Comment (GuillaumeRx) [inline ui/components/multichain/create-eth-account/create-eth-account.js:10]: ?

Comment (hmalik88) [reply ui/components/multichain/create-eth-account/create-eth-account.js:10]: It was causing some test breaking, the import couldn't be read otherwise.

File: app/scripts/metamask-controller.js (hunk 14 above)
Comment (FrederikBolding) [reply app/scripts/metamask-controller.js:2803]: Turns out this is a very real issue. Handling outside of this PR.

File: ui/contexts/metamask-notifications/metamask-notifications.tsx [hunk 16]
```diff
@@ -1,14 +1,12 @@
...
-import type { NotificationServicesController } from '@metamask/notification-services-controller';
 import { useListNotifications } from '../../hooks/metamask-notifications/useNotifications';
 import { useAccountSyncingEffect } from '../../hooks/metamask-notifications/useProfileSyncing';
 import { selectIsProfileSyncingEnabled } from '../../selectors/metamask-notifications/profile-syncing';
 import { selectIsMetamaskNotificationsEnabled } from '../../selectors/metamask-notifications/metamask-notifications';
 import { getUseExternalServices } from '../../selectors';
 import { getIsUnlocked } from '../../ducks/metamask/metamask';
-
-type Notification = NotificationServicesController.Types.INotification;
+import { type Notification } from '../../pages/notifications/notification-components/types/notifications/notifications';
```
Comment (FrederikBolding) [inline ui/contexts/metamask-notifications/metamask-notifications.tsx:9]: Why are we changing these imports?

File: ui/pages/notifications/notifications-list-item.tsx [hunk 17]
```diff
@@ -44,13 +46,23 @@ export function NotificationsListItem({
...
         type: notification.type,
         isRead: notification.isRead,
       },
     ]);
+
+    if (
+      notification.type === TRIGGER_TYPES.SNAP &&
+      !hasProperty(notification.data, 'detailedView')
+    ) {
+      setNotificationTimeout(notification.id);
```
Comment (FrederikBolding) [inline ui/pages/notifications/notifications-list-item.tsx:62]: When is a notification with a detailed view removed?

File: ui/pages/notifications/notification-components/stake/stake.tsx [hunk 18]
```diff
@@ -148,7 +151,7 @@ export const components: NotificationComponent<StakeNotification> = {
       );
     },
     body: {
-      type: 'body_onchain_notification',
+      type: NotificationComponentType.OnChainBody,
```
Comment (FrederikBolding) [inline ui/pages/notifications/notification-components/stake/stake.tsx:154]: In the future it would be nice to split this out in a separated PR and do this refactor before mixing in Snaps changes IMO

File: ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx [hunk 19]
```diff
@@ -29,21 +30,14 @@ export const NotificationDetailsFooter = ({
...
-        <>
-          <footer.ScanLink notification={notification} />
-        </>
+      {footer.type === NotificationComponentType.OnChainFooter && (
+        <footer.ScanLink notification={notification} />
+      )}
+      {footer.type === NotificationComponentType.AnnouncementFooter && (
+        <footer.ExtensionLink notification={notification} />
       )}
-      {footer.type === 'footer_feature_announcement' && (
```
Comment (FrederikBolding) [inline ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx:37]: Is this change intentional?

**Review by FrederikBolding — COMMENTED** [review]
I did not get all the way through yet, but some preliminary comments

File: ui/pages/notifications/notifications-list-item.tsx (hunk 17 above)
Comment (hmalik88) [reply ui/pages/notifications/notifications-list-item.tsx:62]: On dismount

File: ui/pages/notifications/notifications-list-item.stories.tsx [hunk 20]
```diff
@@ -44,7 +44,12 @@ const notificationMocks = {
   RocketPoolStakeCompleted: createMockNotificationRocketPoolStakeCompleted,
   RocketPoolUnStakeCompleted: createMockNotificationRocketPoolUnStakeCompleted,
   FeatureAnnouncement: createMockFeatureAnnouncementRaw,
-  Snap: createMockSnapNotification,
+  Snap: () => {
+    const mock = createMockSnapNotification();
+    // TODO(hmalik88): the mock's origin should be fixed upstream
```
[the remark this answers is not in the document]
Comment (hmalik88) [reply ui/pages/notifications/notifications-list-item.stories.tsx:49]: Not done, going to do it in a later PR

File: ui/pages/notifications/notification-components/stake/stake.tsx (hunk 18 above)
Comment (hmalik88) [reply ui/pages/notifications/notification-components/stake/stake.tsx:154]: Yeah I suppose, did this on Maarten's request lol

File: ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx (hunk 19 above)
Comment (hmalik88) [reply ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx:37]: Yes

Comment (FrederikBolding) [reply ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx:37]: I don't understand then, why are we removing elements from another type of notification?

File: ui/components/multichain/notification-detail-button/notification-detail-button.tsx [hunk 21]
```diff
@@ -46,17 +56,21 @@ export const NotificationDetailButton = ({
...
   };
 
   return (
     <Button
       key={id}
-      href={href}
+      href={!isMetaMaskUrl && href ? href : undefined}
       variant={variant}
-      externalLink={isExternal}
+      externalLink={!isMetaMaskUrl}
```
Comment (FrederikBolding) [inline ui/components/multichain/notification-detail-button/notification-detail-button.tsx:73]: Seems like this would break existing uses of this component with `isExternal=false`?

File: ui/pages/notifications/notification-components/snap/snap.tsx [hunk 22]
```diff
@@ -1,98 +1,161 @@
...
+    return (
+      <NotificationListItemSnap
+        id={notification.id}
+        snapId={notification.data.origin}
+        isRead={notification.isRead}
+        createdAt={new Date(notification.createdAt)}
+        title={{
+          items: [
+            {
+              text: snapsNameGetter(notification.data.origin) ?? 'Unknown Snap',
```
Comment (FrederikBolding) [inline ui/pages/notifications/notification-components/snap/snap.tsx:77]: Is the else case even possible here?

File: app/scripts/metamask-controller.js [hunk 23]
```diff
@@ -1461,12 +1461,19 @@ export default class MetamaskController extends EventEmitter {
         },
         showInAppNotification: {
           method: (origin, args) => {
-            const { message } = args;
+            const { message, title, footerLink, interfaceId } = args;
+
+            const detailedView = {
+              title,
+              ...(footerLink ? { footerLink } : {}),
```
[the remark this answers is not in the document]
Comment (hmalik88) [reply app/scripts/metamask-controller.js:1488]: I'd like to keep the object clean without `undefined`. The performance difference of creating the intermediate object is neglible.

File: app/scripts/metamask-controller.js [hunk 24]
```diff
@@ -1461,12 +1461,19 @@ export default class MetamaskController extends EventEmitter {
...
+              interfaceId,
+            };
 
             const notification = {
               data: {
                 message,
                 origin,
+                ...(interfaceId ? { detailedView } : {}),
```
[the remark this answers is not in the document]
Comment (hmalik88) [reply app/scripts/metamask-controller.js:1496]: See above.

File: ui/pages/notifications/notification-components/snap/snap.tsx
Comment (FrederikBolding) [inline ui/pages/notifications/notification-components/snap/snap.tsx:1]: image
This seems broken. 

Additionally:
- When using a link in a notification with an expanded view you cannot click the link anymore
- If you view an expanded view, go back and into the expanded view again the notification will disappear as you are reading it

File: app/scripts/metamask-controller.js [hunk 25]
```diff
@@ -2922,14 +2929,22 @@ export default class MetamaskController extends EventEmitter {
...
-          showInAppNotification: (origin, args) =>
+          showInAppNotification: (origin, args) => {
+            const { message, title, footerLink } = args;
+            const notificationArgs = {
+              interfaceId: args.content,
+              message,
+              title,
+              footerLink,
+            };
             this.controllerMessenger.call(
```
Comment (FrederikBolding) [inline app/scripts/metamask-controller.js:2940]: We need to return the response here otherwise every call to `snap_notify`  will throw
```
Request for method \"snap_notify\" returned undefined result."
```

File: ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx (hunk 19 above)
Comment (hmalik88) [reply ui/pages/notification-details/notification-details-footer/notification-details-footer.tsx:37]: oh shoot that was NOT intentional, good catch.

File: app/scripts/metamask-controller.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply app/scripts/metamask-controller.js:1488]: This is not a performance worry, but it adds additional complexity that makes it harder to reason about the code.

I'm not necessarily against it, but why do we need `undefined` to not be present here?

[the remark this answers is not in the document]
Comment (hmalik88) [reply app/scripts/metamask-controller.js:1488]: Technically it's not type correct if `footerLink` or `detailedView` is present on the notification object, also I have to do more type-fu to access the `detailedView` property instead of using `hasProperty` as I'm currently doing in some of the components.

File: ui/components/multichain/notification-detail-button/notification-detail-button.tsx (hunk 21 above)
Comment (hmalik88) [reply ui/components/multichain/notification-detail-button/notification-detail-button.tsx]: Yeah, fixed.

File: ui/pages/notifications/notification-components/snap/snap.tsx (hunk 22 above)
Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:77]: Nah, remnants of old notifications team code. Removing it.

File: ui/pages/notifications/notification-components/snap/snap.tsx
Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:1]: 1. Not using markdown so that's why, I'll fix that.
2. This is happening because on dismount the timeout is triggered, I don't have a timer that I'm using to pause if you decided to navigate back. I'll have to add that in if you think from a UX perspective it's better. Technically the view disappearing like this is correct...

Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:1]: > This seems broken.

Are you referring to the below points or something else?

Comment (FrederikBolding) [reply ui/pages/notifications/notification-components/snap/snap.tsx:1]: For 1. I was referring to when not in the expanded view, the link renders correctly but you cannot click it

**Review by eriknson — DISMISSED** [review]
Design lgtm!

File: ui/pages/notifications/notification-components/snap/snap.tsx
Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:1]: That one is kind of tough to get around, I think snap developers might have to design the notification to account for if they provide a detailed view. It'll be better to provide a link inside the JSX content itself.

Comment (FrederikBolding) [reply ui/pages/notifications/notification-components/snap/snap.tsx:1]: Do you have an explanation for why the link isn't even clickable? Seems like we are letting some sort of click bubble up that shouldn't be?

Comment (hmalik88) [reply ui/pages/notifications/notification-components/snap/snap.tsx:1]: Fixed here: [4ec8491](https://github.com/MetaMask/metamask-extension/pull/27407/commits/4ec84917629f56b5d2c30c5c19f2bf575036d626)

File: ui/components/app/snaps/snap-ui-markdown/snap-ui-markdown.js [hunk 26]
```diff
@@ -17,7 +17,7 @@ import {
 } from '../../../component-library';
 import SnapLinkWarning from '../snap-link-warning';
 
-const Paragraph = (props) => (
+export const Paragraph = (props) => (
```
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-ui-markdown/snap-ui-markdown.js:20]: Why is this exported now? Feels like an implementation detail for the component, especially with that name

Comment (hmalik88) [reply ui/components/app/snaps/snap-ui-markdown/snap-ui-markdown.js:20]: Unintended, I was about to do something 😅

File: ui/pages/notification-details/notification-details.tsx [hunk 27]
```diff
@@ -59,7 +61,13 @@ function useEffectOnNotificationView(notificationData?: Notification) {
...
       ]);
     }
-  }, [markNotificationAsRead, notificationData]);
+
+    return () => {
+      if (notificationData?.type === TRIGGER_TYPES.SNAP) {
+        setNotificationTimeout(notificationData.id);
+      }
+    };
+  }, []);
```
Comment (Prithpal-Sooriya) [inline ui/pages/notification-details/notification-details.tsx:70]: Empty dep array is fine here. I will investigate our notifications on why we needed our effect for marking as read (I think we can remove it, but will do so in another PR)

File: ui/pages/notifications/notification-components/erc1155-sent-received/erc1155-sent-received.tsx [hunk 28]
```diff
@@ -189,7 +192,7 @@ export const components: NotificationComponent<ERC1155Notification> = {
     },
   },
   footer: {
-    type: 'footer_onchain_notification',
+    type: NotificationComponentType.OnChainFooter,
```
Comment (Prithpal-Sooriya) [inline ui/pages/notifications/notification-components/erc1155-sent-received/erc1155-sent-received.tsx:195]: Really appreciate this abstraction! I really like the exhaustive discrimination we can now do for OnChain, Feature Announcement, and now Snap Body/Footer/Etc.

File: ui/pages/notifications/notification-components/types/notifications/notifications.ts [hunk 29]
```diff
@@ -52,7 +71,13 @@ export type NotificationComponent<N extends Notification = Notification> = {
...
+    body:
+      | BodyFeatureAnnouncement<N>
+      | BodyOnChainNotification<N>
+      | BodySnapNotification<N>;
   };
-  footer: FooterFeatureAnnouncement<N> | FooterOnChainNotification<N>;
+  footer:
+    | FooterFeatureAnnouncement<N>
+    | FooterOnChainNotification<N>
+    | FooterSnapNotification<N>;
```
Comment (Prithpal-Sooriya) [inline ui/pages/notifications/notification-components/types/notifications/notifications.ts:82]: Nice. I think our team might do another refactor in the future to better group these unions.
E.g. it is unlikely that you would have a FeatureAnnonucement Body, but a Snap Footer.

IDK why we didn't group the footer in the `details` section 😛 

Noted and will work on this later.

**Review by Prithpal-Sooriya — DISMISSED** [review]

**Review by Prithpal-Sooriya — APPROVED** [review]

**Review by GuillaumeRx — APPROVED** [review]

**Review by matthewwalsh0 — APPROVED** [review]

File: app/scripts/metamask-controller.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (GuillaumeRx) [reply app/scripts/metamask-controller.js:1488]: I'll consider this as a nit that we can fix after if we really want. This would just improve code readability :)

**Review by darkwing — APPROVED** [review]
